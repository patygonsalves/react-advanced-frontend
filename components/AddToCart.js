import React from 'react'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'

import { CURRENT_USER_QUERY } from './User'

const ADD_TO_CART_MUTATION = gql`
  mutation addToCart($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`

function AddToCartItem({
  id
}) {
  return (
    <Mutation
      mutation={ADD_TO_CART_MUTATION}
      variables={{ id }}
      refetchQueries={[{ query: CURRENT_USER_QUERY }]}
    >
      {(addToCart, { loading }) => <button disabled={loading} onClick={addToCart}>Add{loading && 'ing'} to cart</button>}
    </Mutation>
  )
}

export default AddToCartItem
export { ADD_TO_CART_MUTATION }