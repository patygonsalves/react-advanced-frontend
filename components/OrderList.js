import React from 'react'

import { Query } from 'react-apollo'
import gql from 'graphql-tag'
import { formatDistance } from 'date-fns'
import Link from 'next/link'
import styled from 'styled-components'

import formatMoney from '../lib/formatMoney'
import Error from './ErrorMessage'

import OrderItemStyles from './styles/OrderItemStyles'

const USER_ORDER_QUERY = gql`
  query USER_ORDER_QUERY {
    orders(orderBy: createdAt_ASC) {
      id,
      total,
      createdAt,
      items {
        id
        title
        description
        price
        image
        quantity
      }
    }
  }
`

const OrderUl = styled.ul`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`

function OrderList() {
  return (
    <Query query={USER_ORDER_QUERY}>
      {( { data: { orders }, error, loading } ) => {
        if(error) return <Error error={error} />
        if(loading) return <p>Loading...</p>
        return (
          <div>
            <h2>You have {orders.length} orders</h2>
            <OrderUl>
              {orders.map(order => (
                <OrderItemStyles key={order.id}>
                  <Link
                    href={{
                      pathname: '/order',
                      query: { id: order.id },
                    }}
                  >
                    <a>
                      <div className='order-meta'>
                        <p>{order.items.reduce((a, b) => a + b.quantity, 0)} Items</p>
                        <p>{order.items.length} Products</p>
                        <p>{formatDistance(order.createdAt, new Date())}</p>
                        <p>{formatMoney(order.total)}</p>
                      </div>
                      <div className='images'>
                        {order.items.map(item => (
                          <img key={item.id} src={item.image} alt={item.title} />
                        ))}
                      </div>
                    </a>
                  </Link>
                </OrderItemStyles>
              ))}
            </OrderUl>
          </div>
        )
      }}
    </Query>
  )
}


export default OrderList