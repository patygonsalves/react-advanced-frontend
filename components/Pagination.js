import React from 'react'
import { Query } from 'react-apollo'
import gql from 'graphql-tag'

import Head from 'next/head'
import Link from 'next/link'

import { perPage } from '../config'

import PaginationStyles from './styles/PaginationStyles'

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`

const Pagination = props => (
  <Query query={PAGINATION_QUERY}>
    {({ data, loading, error }) => {
      if (loading) return <p>Loading...</p>
      const count = data.itemsConnection.aggregate.count
      const pages = Math.ceil(count / perPage)
      return (
        <PaginationStyles data-test='pagination'>
          <Head>
            <title>Sick Fits! {props.page} of {pages}</title>
          </Head>
          <Link prefetch href={{
            pathname: 'items',
            query: { page: props.page -1 }
          }}>
            <a className='prev' aria-disabled={props.page <= 1}>Prev</a>
          </Link>
          <p>{props.page} of <span className='totalPages'>{pages}</span></p>
          <p>{count} Items Total</p>
          <Link prefetch href={{
            pathname: 'items',
            query: { page: props.page + 1 }
          }}>
            <a className='next' aria-disabled={props.page >= pages}>Next</a>
          </Link>
        </PaginationStyles>
      )
    }}
  </Query>
)

export default Pagination
export { PAGINATION_QUERY }
