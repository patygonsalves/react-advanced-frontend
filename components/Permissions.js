import React, { useState } from 'react'
import { Query, Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import PropTypes from 'prop-types'

import Table from './styles/Table'
import SickButton from './styles/SickButton'
import Error from './ErrorMessage'

const ALL_USERS_QUERY = gql`
  query {
    users {
      id
      name
      email
      permissions
    }
  }
`

const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation UPDATE_PERMISSIONS_MUTATION($permissions: [Permission], $userId: ID!) {
    updatePermissions(permissions: $permissions, userId: $userId){
      id,
      permissions
      name
      email
    }
  }
`

const possiblePermission = [
  'ADMIN',
  'USER',
  'ITEMCREATE',
  'ITEMUPDATE',
  'ITEMDELETE',
  'PERMISSIONUPDATE',
]

const Permissions = (props) => (
  <Query query={ALL_USERS_QUERY}>
    {({ data, loading, error }) => (
      <div>
        <Error error={error} />
        <div>
          <h2>Manage Permissions</h2>
          <Table>
            <thead>
              <tr>
                <th>Name</th>
                <th>Email</th>
                {possiblePermission.map(p => <th key={p}>{p}</th>)}
                <th>+</th>
              </tr>
            </thead>
            <tbody>
              {data.users.map(u => <UserPermissions key={u.id} user={u} />)}
            </tbody>
          </Table>
        </div>
      </div>
    )}
  </Query>
)

function UserPermissions({
  user
}) {

  const [permission, setPermission] = useState(user.permissions)

  const handlePermissionChange = e => {
    const checkbox = e.target

    let updatePermissions = [...permission]
    if(checkbox.checked) {
      updatePermissions.push(checkbox.value)
    } else {
      updatePermissions = updatePermissions.filter(u => u !== checkbox.value)
    }

    setPermission(updatePermissions)
  }

  return(
    <Mutation
      mutation={UPDATE_PERMISSIONS_MUTATION}
      variables={{
        permissions: permission,
        userId: user.id,
      }}
    >
      {(updatePermissions, { loading, error }) => (
        <>
          {error && <tr><td colspan='8'><Error error={error} /></td></tr>}
          <tr>
            <td>{user.name}</td>
            <td>{user.email}</td>
            {possiblePermission.map(p => (
              <td key={p}>
                <label htmlFor={`${user.id}-permission-${p}`}>
                  <input
                    id={`${user.id}-permission-${p}`}
                    type='checkbox'
                    checked={permission.includes(p)}
                    value={p}
                    onChange={handlePermissionChange}
                  />
                </label>
              </td>
            ))}
            <td>
              <SickButton
                type='button'
                disabled={loading}
                onClick={updatePermissions}
              >
                Updat{loading ? 'ing' : 'e'}
              </SickButton>
            </td>
        </tr>
       </>
      )}
    </Mutation>
  )
}

UserPermissions.propTypes = {
  user: PropTypes.shape({
    name: PropTypes.string,
    email: PropTypes.string,
    id: PropTypes.string,
    permissions: PropTypes.array,
  }).isRequired
}

export default Permissions