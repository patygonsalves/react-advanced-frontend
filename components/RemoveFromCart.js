import React, { Component } from 'react'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import styled from 'styled-components'
import PropTypes from 'prop-types'

import { CURRENT_USER_QUERY } from './User'

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover{
    color: ${props => props.theme.red};
    cursor: pointer;
  }
`
const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`

function RemoveFromCart({
  id,
}) {


  const update = (cache, payload) => {
    // 1. read the cache
    const data = cache.readQuery({ query: CURRENT_USER_QUERY })
    // 2. remove that item from the cart
    const cartItemId = payload.data.removeFromCart.id
    data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId)
    // 3. write it back to the cache
    cache.writeQuery({ query: CURRENT_USER_QUERY, data })
  }

  return (
    <Mutation
      mutation={REMOVE_FROM_CART_MUTATION}
      variables={{ id }}
      update={update}
      optimisticResponse={{
        __typename: 'Mutation',
        removeFromCart: {
          __typename: 'CartItem',
          id,
        },
      }}
    >
      {(removeFromCart, { loading, error}) => (
        <BigButton
          onClick={() => {
            removeFromCart().catch(err => alert(err.message))
          }}
          title='Delete item'
          disabled={loading}
        >&times;</BigButton>
      )}
    </Mutation>  )
}

RemoveFromCart.propTypes = {
  id: PropTypes.string.isRequired,
}

export default RemoveFromCart
export { REMOVE_FROM_CART_MUTATION }