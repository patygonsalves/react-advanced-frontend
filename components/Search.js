import React, { useState, useEffect } from 'react'
import Router from 'next/router'
import { ApolloConsumer } from 'react-apollo'
import gql from 'graphql-tag'
import Downshift, { resetIdCounter } from 'downshift'
import debounce from 'lodash.debounce'

import { DropDown, DropDownItem, SearchStyles } from './styles/DropDown'

const SEARCH_ITEMS_QUERY = gql`
  query SEARCH_ITEMS_QUERY($searchTerm: String!) {
    items(where: {
      OR: [
        { title_contains: $searchTerm, },
        { description_contains: $searchTerm, },
      ]
    }) {
      id
      image
      title
    }
  }
`

function AutoComplete({

}){
  const [items, setItems] = useState([])
  const [loading, setLoading] = useState(true)

  useEffect(() => {
    setLoading(false)
  }, [])

  const routeToItem = (item) => {
    Router.push({
      pathname: '/item',
      query: {
        id: item.id,
      }
    })
  }

  const handleChange = debounce(async (e, client) => {
    setLoading(true)
    
    const res = await client.query({
      query: SEARCH_ITEMS_QUERY,
      variables: { searchTerm: e.target.value }
    })

    setLoading(false)
    console.log(res.data.items)
    setItems(res.data.items)
    /*setItems(res.data.items)
    console.log(res.data.items)
    console.log(items)
    console.log(loading)*/
  }, 350)

  resetIdCounter()

  console.log(loading)
  console.log(items)

  return (
    <SearchStyles>
      <Downshift onChange={routeToItem} itemToString={item => (item === null ? '' : item.title)}>
        {({ getInputProps, getItemProps, isOpen, inputValue, highlightedIndex }) => (
          <div>
            <ApolloConsumer>
              {client => (
                <input
                  {...getInputProps({
                    type: 'search',
                    placeholder: 'search for an item',
                    id: 'search',
                    className: loading ? 'loading' : '',
                    onChange: e => {
                      e.persist()
                      handleChange(e, client)
                    }
                  })}
                />
              )}
            </ApolloConsumer>
            {isOpen && (
              <DropDown>
                {items.map((item, index) => {
                  return (
                    <DropDownItem
                      key={item.id}
                      highlighted={index === highlightedIndex}
                      {...getItemProps({ item })}
                    >
                      <img width='50' alt={item.title} src={item.image} />
                      {item.title}
                    </DropDownItem>
                  )
                })}
                {!items.length && !loading &&
                  <DropDownItem>Nothing found {inputValue}</DropDownItem>
                }
              </DropDown>
            )}
          </div>
        )}
      </Downshift>
    </SearchStyles>
  )
}

export default AutoComplete