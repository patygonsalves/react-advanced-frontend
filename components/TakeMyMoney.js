import React from 'react'
import StripeCheckout from 'react-stripe-checkout'
import { Mutation } from 'react-apollo'
import Router from 'next/router'
import NProgress from 'nprogress'
import gql from 'graphql-tag'

import calcTotalPrice from '../lib/calcTotalPrice'
import User, { CURRENT_USER_QUERY } from './User'


const CREATE_ORDER_MUTATION = gql`
  mutation createOrder($token: String!) {
    createOrder(token: $token) {
      id
      charge
      total
      items {
        id
        title
      }
    }
  }
`


function TakeMyMoney({
  children,
}) {

  const totalItems = (cart) => {
    return cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)
  }

  const onToken = async (res, createOrder) => {
    NProgress.start()

    const order = await createOrder({
      variables: {
        token: res.id
      }
    }).catch(err => {
      alert(err.message)
    })

    Router.push({
      pathname: '/order',
      query: {
        id: order.data.createOrder.id
      }
    })

  }

  return (
    <User>
      {({ data: { me }, loading }) => {
        if (loading) return null
        return (
          <Mutation
            mutation={CREATE_ORDER_MUTATION}
            refetchQueries={[{ query: CURRENT_USER_QUERY }]}
          >
            {(createOrder) => (
              <StripeCheckout
                amount={calcTotalPrice(me.cart)}
                name='Sick Fits'
                description={`Order of ${totalItems(me.cart)} items`}
                image={me.cart.lenght && me.cart[0].item && me.cart[0].item.image}
                stripeKey='pk_test_Vtknn6vSdcZWSG2JWvEiWSqC'
                currency='USD'
                email={me.email}
                token={res => onToken(res, createOrder)} 
              >
                <p>{children}</p>
              </StripeCheckout>
            )}
          </Mutation>
        )}
      }
    </User>
  )
}

export default TakeMyMoney
export { CREATE_ORDER_MUTATION }
