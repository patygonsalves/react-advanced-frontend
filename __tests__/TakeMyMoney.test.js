import wait from 'waait'
import { MockedProvider } from 'react-apollo/test-utils'
import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import NProgress from 'nprogress'
import Router from 'next/router'
import TakeMyMoney, { CREATE_ORDER_MUTATION } from '../components/TakeMyMoney'
import { CURRENT_USER_QUERY } from '../components/User'
import { fakeUser, fakeCartItem } from '../lib/testUtils'
import { ApolloConsumer } from 'react-apollo'

Router.router = { push() {} }

const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem()],
        }
      }
    }
  }
]

describe('<TakeMyMoney />', () => {
  it('renders and marches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TakeMyMoney />
      </MockedProvider>
    )

    await wait()
    wrapper.update()
    const checkoutButton = wrapper.find('ReactStripeCheckout')
    expect(toJSON(checkoutButton)).toMatchSnapshot()
  })

  /*it('creates an order ontoken', async () => {
    const createOrderMock = jest.fn().mockResolvedValue({
      data: {
        createOrder: {
          id: 'xyz789'
        }
      }
    })

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TakeMyMoney />
      </MockedProvider>
    )
    const component = wrapper.find('TakeMyMoney').instance()
    component.onToken({
      id: 'abc123'
    }, createOrderMock)
    expect(createOrderMock).toHaveBeenCalled()
  })*/

  /*it('turns the progress bar on', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TakeMyMoney />
      </MockedProvider>
    )

    await wait()
    wrapper.update()
    NProgress.start = jest.fn()
    const createOrderMock = jest.fn().mockResolvedValue({
      data: { createOrder: { id: 'xyz789' } }
    })

    const component = wrapper.find('TakeMyMoney').instance()
    component.onToken({ id: 'abc123' }, createOrderMock)
    expect(NProgress.start).toHaveBeenCalled()
  })*/

})
