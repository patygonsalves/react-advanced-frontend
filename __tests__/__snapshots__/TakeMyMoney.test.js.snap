// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TakeMyMoney /> renders and marches snapshot 1`] = `
<ReactStripeCheckout
  ComponentClass="span"
  amount={15000}
  className="StripeCheckout"
  currency="USD"
  description="Order of 3 items"
  email="Delmer.Smith@yahoo.com"
  label="Pay With Card"
  locale="auto"
  name="Sick Fits"
  reconfigureOnUpdate={false}
  stripeKey="pk_test_Vtknn6vSdcZWSG2JWvEiWSqC"
  token={[Function]}
  triggerEvent="onClick"
>
  <span
    onClick={[Function]}
  >
    <p />
  </span>
</ReactStripeCheckout>
`;
